From 99242255fd8c63d3a18e748f7b6ed79d410aa24c Mon Sep 17 00:00:00 2001
From: Jin Wei <wei.a.jin@intel.com>
Date: Mon, 18 Mar 2013 13:10:17 +0800
Subject: [PATCH] restore legacy vibrator HAL.

Revert "Remove vibrator AOSP implementation from libhardware_legacy"
This reverts commit 65005252f2859234d4fa89bfc6f72dc6f56e0f22.

Change-Id: Ia8d66c093fe40a2a9110a622692022fba9813130
Signed-off-by: Jin Wei <wei.a.jin@intel.com>
---
 Android.mk          |    2 +-
 vibrator/Android.mk |    4 +++
 vibrator/vibrator.c |   75 +++++++++++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 80 insertions(+), 1 deletion(-)
 create mode 100644 vibrator/Android.mk
 create mode 100644 vibrator/vibrator.c

diff --git a/Android.mk b/Android.mk
index e2b35c6..1faf6e2 100644
--- a/Android.mk
+++ b/Android.mk
@@ -1,7 +1,7 @@
 # Copyright 2006 The Android Open Source Project
 
 # Setting LOCAL_PATH will mess up all-subdir-makefiles, so do it beforehand.
-legacy_modules := power uevent wifi qemu qemu_tracing
+legacy_modules := power uevent vibrator wifi qemu qemu_tracing
 
 SAVE_MAKEFILES := $(call all-named-subdir-makefiles,$(legacy_modules))
 LEGACY_AUDIO_MAKEFILES := $(call all-named-subdir-makefiles,audio)
diff --git a/vibrator/Android.mk b/vibrator/Android.mk
new file mode 100644
index 0000000..6681f84
--- /dev/null
+++ b/vibrator/Android.mk
@@ -0,0 +1,4 @@
+# Copyright 2006 The Android Open Source Project
+
+LOCAL_SRC_FILES += vibrator/vibrator.c
+
diff --git a/vibrator/vibrator.c b/vibrator/vibrator.c
new file mode 100644
index 0000000..f946ce1
--- /dev/null
+++ b/vibrator/vibrator.c
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2008 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <hardware_legacy/vibrator.h>
+#include "qemu.h"
+
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#define THE_DEVICE "/sys/class/timed_output/vibrator/enable"
+
+int vibrator_exists()
+{
+    int fd;
+
+#ifdef QEMU_HARDWARE
+    if (qemu_check()) {
+        return 1;
+    }
+#endif
+
+    fd = open(THE_DEVICE, O_RDWR);
+    if(fd < 0)
+        return 0;
+    close(fd);
+    return 1;
+}
+
+static int sendit(int timeout_ms)
+{
+    int nwr, ret, fd;
+    char value[20];
+
+#ifdef QEMU_HARDWARE
+    if (qemu_check()) {
+        return qemu_control_command( "vibrator:%d", timeout_ms );
+    }
+#endif
+
+    fd = open(THE_DEVICE, O_RDWR);
+    if(fd < 0)
+        return errno;
+
+    nwr = sprintf(value, "%d\n", timeout_ms);
+    ret = write(fd, value, nwr);
+
+    close(fd);
+
+    return (ret == nwr) ? 0 : -1;
+}
+
+int vibrator_on(int timeout_ms)
+{
+    /* constant on, up to maximum allowed time */
+    return sendit(timeout_ms);
+}
+
+int vibrator_off()
+{
+    return sendit(0);
+}
-- 
1.7.9.5

